### PR TITLE
Feature/inject into storyboard views

### PR DESCRIPTION
**Pull request by Nekitosss**
1. Add injecting into UIViews created by UIStoryboard
2. Add injecting into UITableViewCells and UICollectionViewCells created by UITableView and UICollectionView. Note: Programmatically created UITableViews (inside UITableViewController too) and UICollectionViews (inside UICollectionViewControllers too) not injecting into its cells. Will be implemented in future

### CODE REVIEW

File: Tests/DITranquillityTest/DITranquillityTests_Storyboard.swift [hunk 1]
```diff
@@ -9,7 +9,22 @@
 import XCTest
 import DITranquillity
 
+class TestView: UIView {
```
[the remark this answers is not in the document]
Comment (Nekitosss) [reply Tests/DITranquillityTest/DITranquillityTests_Storyboard.swift]: Storyboard should see that classes. Could not set private access modificator.

File: Tests/DITranquillityTest/DITranquillityTests_Storyboard.swift [hunk 2]
```diff
@@ -371,4 +386,129 @@ class DITranquillityTests_Storyboard: XCTestCase {
...
+    
+    let testDataSource = TableViewTestDataSource()
+    tableView.dataSource = testDataSource
+    tableView.dataSource!.tableView(tableView, cellForRowAt: .init(row: 0, section: 0))
+    
+    XCTAssertTrue(testDataSource.result)
+  }
+  
+  
+  func test16_InjectToCellOnCollectionVIewController() {
```
[the remark this answers is not in the document]
Comment (Nekitosss) [reply Tests/DITranquillityTest/DITranquillityTests_Storyboard.swift]: Fixed.

File: Sources/Storyboard/StoryboardResolver.swift [hunk 3]
```diff
@@ -23,12 +23,18 @@ final class StoryboardResolver {
...
-
+    
     for childVC in viewController.childViewControllers {
+      setStoryboardResolver(to: childVC)
       inject(into: childVC)
     }
   }
+  
+  private func setStoryboardResolver(to object: AnyObject) {
+    objc_setAssociatedObject(object, _DIStoryboardBase.resolver_UNIQUE_VC_KEY(), self.container.nsDIContainer, .OBJC_ASSOCIATION_RETAIN_NONATOMIC)
```
[the remark this answers is not in the document]
Comment (Nekitosss) [reply Sources/Storyboard/StoryboardResolver.swift]: Fixed.

File: Sources/Core/Public/NSDIContainerProtocol.h [hunk 4]
```diff
@@ -0,0 +1,14 @@
+//
+//  NSDIContainerProtocol.h
+//  DITranquillity
+//
+//  Created by Admin on 20/04/2018.
+//  Copyright © 2018 Alexander Ivlev. All rights reserved.
+//
+
+
+@protocol NSDIContainerProtocol
```
[the remark this answers is not in the document]
Comment (Nekitosss) [reply Sources/Core/Public/NSDIContainerProtocol.h]: Fixed.

File: Sources/Core/Public/DIContainer.swift [hunk 5]
```diff
@@ -21,6 +21,20 @@ public prefix func *<T>(container: DIContainer) -> T {
   return container.resolve()
 }
 
+public class NSDIContainer: NSObject, NSDIContainerProtocol {
```
[the remark this answers is not in the document]
Comment (Nekitosss) [reply Sources/Core/Public/DIContainer.swift]: Could not be internal cause of usage in objc reason.

File: Sources/Core/Public/DIContainer.swift [hunk 6]
```diff
@@ -33,6 +47,9 @@ public final class DIContainer {
   internal let componentContainer = ComponentContainer()
   internal let bundleContainer = BundleContainer()
   internal private(set) var resolver: Resolver!
+  internal var nsDIContainer: NSDIContainer {
```
[the remark this answers is not in the document]
Comment (Nekitosss) [reply Sources/Core/Public/DIContainer.swift]: Moved to resolver and renamed to NSResolver

File: Sources/Core/Private/Resolver.swift [hunk 7]
```diff
@@ -225,13 +251,14 @@ class Resolver {
     func initialObject() -> Any? {
       if let obj = usingObject {
         log(.verbose, msg: "Use object: \(obj)")
+        setNSResolverIfNeeded(to: obj)
```
Comment (ivlevAstef) [inline Sources/Core/Private/Resolver.swift]: move to a higher level - after call initialObject

File: Sources/Core/Private/Resolver.swift [hunk 8]
```diff
@@ -225,13 +251,14 @@ class Resolver {
...
         return obj
       }
       
-      
       if let signature = component.initial {
         let obj = use(signature: signature, usingObject: nil)
         log(.verbose, msg: "Create object: \(String(describing: obj))")
+        setNSResolverIfNeeded(to: obj)
```
Comment (ivlevAstef) [inline Sources/Core/Private/Resolver.swift]: move to a higher level - after call initialObject

File: Sources/Core/Private/Resolver.swift [hunk 9]
```diff
@@ -6,9 +6,33 @@
...
+
 class Resolver {
+  
+  private func setNSResolver(to object: Any) {
+    objc_setAssociatedObject(object, NSResolver.getResolverUniqueAssociatedKey(), self.nsResolver, .OBJC_ASSOCIATION_RETAIN_NONATOMIC)
+  }
+  
   init(container: DIContainer) {
     self.container = container // onowned
+    self.nsResolver = NSResolver(resolver: self)
```
Comment (ivlevAstef) [inline Sources/Core/Private/Resolver.swift]: remove

File: Sources/Core/Private/Resolver.swift [hunk 10]
```diff
@@ -6,9 +6,33 @@
...
+  public static func getResolverUniqueAssociatedKey() -> UnsafeRawPointer {
+    return withUnsafePointer(to: &resolverAssociatedHandle) { .init($0) }
+  }
+  
+  public func inject(into object: Any) {
+    resolver.injection(obj: object)
+  }
+  
+  private static var resolverAssociatedHandle: UInt8 = 0
+  private unowned let resolver: Resolver
```
Comment (ivlevAstef) [inline Sources/Core/Private/Resolver.swift]: unowned? weak or strong.

File: Sources/Core/Private/Resolver.swift [hunk 11]
```diff
@@ -269,6 +296,12 @@ class Resolver {
       return signature.call(objParameters)
     }
     
+    func setNSResolverIfNeeded(to obj: Any?) {
```
Comment (ivlevAstef) [inline Sources/Core/Private/Resolver.swift]: always obj not optional.

File: Sources/Core/Private/Resolver.swift [hunk 12]
```diff
@@ -269,6 +296,12 @@ class Resolver {
...
+      if let anObj = obj, component.injectToSubviews {
+        self.setNSResolver(to: anObj)
```
Comment (ivlevAstef) [inline Sources/Core/Private/Resolver.swift]: `NSResolver(resolver: self).add(on: anObj)` - each object has its own

File: Sources/Core/Private/Resolver.swift [hunk 13]
```diff
@@ -298,6 +331,7 @@ class Resolver {
   }
  
   private unowned let container: DIContainer
+  private var nsResolver: NSResolverProtocol!
```
Comment (ivlevAstef) [inline Sources/Core/Private/Resolver.swift]: remove

File: Sources/Storyboard/iOS-tvOS/UICollectionView+Swizzling.m [hunk 14]
```diff
@@ -0,0 +1,64 @@
...
+    static dispatch_once_t onceToken;
+    
+    dispatch_once(&onceToken, ^{
+        [self swizzleOriginalSelector:@selector(dequeueReusableCellWithReuseIdentifier:forIndexPath:)
+                     swizzledSelector:@selector(di_dequeueReusableCellWithReuseIdentifier:forIndexPath:)];
+    });
+}
+
+
++ (void)swizzleOriginalSelector:(SEL)originalSelector swizzledSelector:(SEL)swizzledSelector {
```
[the remark this answers is not in the document]
Comment (ivlevAstef) [reply Sources/Storyboard/iOS-tvOS/UICollectionView+Swizzling.m]: !

File: Sources/Storyboard/iOS-tvOS/UIViewController+Swizzling.m [hunk 15]
```diff
@@ -0,0 +1,72 @@
...
+  [self passContainerInto:self.view container:[self nsResolver]];
+}
+
+- (void)passContainerInto:(UIView *)view container:(NSResolver *)nsResolver {
+  [nsResolver injectInto:view];
+  if ([view isKindOfClass:[UITableView class]] || [view isKindOfClass:[UICollectionView class]]) {
+    objc_setAssociatedObject(view, [NSResolver getResolverUniqueAssociatedKey], nsResolver, OBJC_ASSOCIATION_RETAIN_NONATOMIC);
+  }
+  
+  for (int i = 0; i < [view.subviews count]; i++) {
```
Comment (ivlevAstef) [inline Sources/Storyboard/iOS-tvOS/UIViewController+Swizzling.m]: for(UIView *subview in view.subviews)

File: Tests/DITranquillityTest/DITranquillityTests_Storyboard.swift (hunk 1 above)
[the remark this answers is not in the document]
Comment (ivlevAstef) [reply Tests/DITranquillityTest/DITranquillityTests_Storyboard.swift]: :(

File: Sources/Core/Private/Resolver.swift (hunk 7 above)
Comment (Nekitosss) [reply Sources/Core/Private/Resolver.swift]: Fixed.

File: Sources/Core/Private/Resolver.swift (hunk 8 above)
Comment (Nekitosss) [reply Sources/Core/Private/Resolver.swift]: Fixed.

File: Sources/Core/Private/Resolver.swift (hunk 9 above)
Comment (Nekitosss) [reply Sources/Core/Private/Resolver.swift]: Removed.

File: Sources/Core/Private/Resolver.swift (hunk 10 above)
Comment (Nekitosss) [reply Sources/Core/Private/Resolver.swift]: Change to strong.

File: Sources/Core/Private/Resolver.swift (hunk 11 above)
Comment (Nekitosss) [reply Sources/Core/Private/Resolver.swift]: Fixed.

File: Sources/Core/Private/Resolver.swift (hunk 12 above)
Comment (Nekitosss) [reply Sources/Core/Private/Resolver.swift]: Fixed. Create unique NSResolver instance for each object.

File: Sources/Core/Private/Resolver.swift (hunk 13 above)
Comment (Nekitosss) [reply Sources/Core/Private/Resolver.swift]: Removed.

File: Sources/Storyboard/iOS-tvOS/UIViewController+Swizzling.m (hunk 15 above)
Comment (Nekitosss) [reply Sources/Storyboard/iOS-tvOS/UIViewController+Swizzling.m]: Fixed.

File: Sources/Core/Private/NSResolverProtocol.h [hunk 16]
```diff
@@ -0,0 +1,16 @@
+//
+//  NSDIContainerProtocol.h
+//  DITranquillity
+//
+//  Created by Admin on 20/04/2018.
```
Comment (ivlevAstef) [inline Sources/Core/Private/NSResolverProtocol.h]: ...

File: Sources/Storyboard/iOS-tvOS/UITableView+Swizzling.m [hunk 17]
```diff
@@ -0,0 +1,40 @@
+//
+//  UIStoryboard+Swizzling.m
+//  DITranquillity
+//
+//  Created by Alexander Ivlev on 27/08/2017.
+//  Copyright © 2017 Alexander Ivlev. All rights reserved.
```
Comment (ivlevAstef) [inline Sources/Storyboard/iOS-tvOS/UITableView+Swizzling.m]: !

File: Sources/Storyboard/iOS-tvOS/UICollectionView+Swizzling.m [hunk 18]
```diff
@@ -0,0 +1,40 @@
+//
+//  UIStoryboard+Swizzling.m
+//  DITranquillity
+//
+//  Created by Alexander Ivlev on 27/08/2017.
+//  Copyright © 2017 Alexander Ivlev. All rights reserved.
```
Comment (ivlevAstef) [inline Sources/Storyboard/iOS-tvOS/UICollectionView+Swizzling.m]: !

File: Sources/Storyboard/iOS-tvOS/NSObject+Swizzling.m [hunk 19]
```diff
@@ -0,0 +1,41 @@
+//
+//  NSObject+Swizzling.m
+//  DITranquillity
+//
+//  Created by Admin on 27/08/2017.
+//  Copyright © 2017 Alexander Ivlev. All rights reserved.
```
Comment (ivlevAstef) [inline Sources/Storyboard/iOS-tvOS/NSObject+Swizzling.m]: !

File: Sources/Storyboard/DIComponentBuilder+SubviewsInject.swift [hunk 20]
```diff
@@ -0,0 +1,22 @@
+//
+//  DIComponentBuilder+SubviewsInject.swift
+//  DITranquillity
+//
+//  Created by Nikita on 23.04.2018.
+//  Copyright © 2018 Alexander Ivlev. All rights reserved.
```
Comment (ivlevAstef) [inline Sources/Storyboard/DIComponentBuilder+SubviewsInject.swift]: !

File: Sources/Storyboard/iOS-tvOS/UIViewController+Swizzling.m [hunk 21]
```diff
@@ -0,0 +1,48 @@
+//
+//  UIStoryboard+Swizzling.m
+//  DITranquillity
+//
+//  Created by Alexander Ivlev on 27/08/2017.
+//  Copyright © 2017 Alexander Ivlev. All rights reserved.
```
Comment (ivlevAstef) [inline Sources/Storyboard/iOS-tvOS/UIViewController+Swizzling.m]: !

**Pull request merged**